Tools/uorb_graph/create.py

```diff
@@ -68,189 +70,216 @@ def __init__(self, is_publication, topic_blacklist, orb_pub_sub_regexes, special
-
-    ]
-        special_cases_sub = [(a, re.compile(b), re.compile(c) if c is not None else None, re.compile(d))
-                                   for a,b,c,d in special_cases_sub]
-
-        self._subscriptions = PubSub(False, topic_blacklist,
-                [r"\borb_subscribe(_multi|)\b\(([^,)]+)"],
-                special_cases_sub)
-
-
-        special_cases_pub = [
-    ('replay', r'Replay\.cpp$', None, r'^sub\.orb_meta$'),
-
-    ('uavcan', r'sensors/.*\.cpp$', None, r'^_orb_topic$'),
-    ]
-        special_cases_pub = [(a, re.compile(b), re.compile(c) if c is not None else None, re.compile(d))
-                                   for a,b,c,d in special_cases_pub]
-        self._publications = PubSub(True, topic_blacklist,
-                [r"\borb_advertise(_multi|_queue|_multi_queue|)\b\(([^,)]+)",
-                 r"\borb_publish_auto()\b\(([^,)]+)"],
-                special_cases_pub)
-
-
-    def _get_current_module(self):
-        if len(self._current_module) == 0:
+        # note: the source-file-string is pre-processed to remove whitespace -- regexes should ignore whitespace
+        # note:  the regexes should have at least 3 capture groups '()'; otherwise they break downstream code
+        capture_cases_sub = [r"\borb_subscribe(?:_multi|)\b\s*\(\s*(ORB_ID)\s*\(\s*(\w+)",
+                             r"(?:uORB::)Subscription(?:Interval|)\s+\w+\s*[\{\(]\s*(ORB_ID)\s*\(\s*(\w+)",
+                             r"(?:uORB::)Subscription(?:Data|MultiArray|Blocking|)\s*(<)\s*(\w+)",
+                             r"(?:uORB::)SubscriptionCallbackWorkItem\s+\w+\s*\{\s*this,\s*(ORB_ID)\((\w+)",
+                            ]
+        self._subscriptions = Subscriptions( self._topic_blacklist, capture_cases_sub)
+
+        # note: the source-file-string is pre-processed to remove whitespace -- regexes should ignore whitespace
+        # note: the regexes should have at least 3 capture groups '()'; otherwise they break downstream code
+        capture_cases_pub = [r"(?:uORB::)Publication(?:Data|Multi|)\s*(<)(\w+)>",
+                             r"orb_advertise(?:_multi|_queue|_multi_queue|)\s*\(\s*(ORB_ID)\s*\(\s*(\w+)",
+                             r"orb_publish(?:_auto|)\s*\(\s*(ORB_ID)\s*\(\s*(\w+)",
+                             r"(?:uORB::)Publication(?:Data|Multi|)\s*<\w+>\s+\w+\s*[\(\{](ORB_ID)\((\w+)"
+                            ]
+        self._publications = Publications( self._topic_blacklist, capture_cases_pub)
+
+        # note: the source-file-string is pre-processed to remove whitespace -- regexes should ignore whitespace
+        # note: the regexes should have at least 3 capture groups '()'; otherwise they break downstream code
+        capture_cases_ambiguous = [ r"orb_copy\s*\(\s*(ORB_ID)\s*\(\s*(\w+)",
+                                    r"(?:uORB::)Subscription\s+\w+\s*(\[)\s*\w+\s*\]()",
+                                    r"(ORB_ID)\s*\(\s*(\w+)",
```

Comment:
This should handle `ORB_ID::` as well

Reply:
I'm not aware of that case...  do you have an example?

Reply:
There's an increasing amount, like:
- https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/navigator/mission.h#L246
- https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/airspeed_selector/airspeed_selector_main.cpp#L122

Reply:
Added as test cases 8 & 9, respectively.